### DISCUSSION TITLE
Advice wanted: dealing with multiple kitty responses

### DISCUSSION THREAD

**Discussion by anicolao** [kovidgoyal/kitty#4624] [Q&A] [answered]
My daily use case for kitty is as my shell into a shared environment where the command running inside each kitty os window is dtach which then displays the underlying console program (usually bash in a remote VM).

The point of this setup is to make it trivial to share shells with other users and be able to work together collaboratively in these shells.

The problem is that I'm not super familiar with the ANSI escape sequence protocol and there seem to be some escape sequences that are actually a protocol of query/answer between the remote program and the terminal emulator. For example to make vim work reliably in the shared terminal setup, I have put this in my .vimrc:
```
" these are all the terminal request/response codes that                                              
" cause a race on startup that might put tons of junk                                                 
" in the vim session. Just disable all query/response                                                 
" codes so that vim starts up cleanly.                                                                
set t_RV=                                                                                             
set t_u7=                                                                                             
set t_RF=                                                                                             
set t_RB=                                                                                             
set t_RC=                                                                                             
set t_RS=  
```
What I am looking for is advice on how to put together a more general solution, because recently I have been using glow to render markdown to the terminal and that also suffers from the same problem no doubt due to some terminal querying done by some library glow depends upon.

Advice appreciated.

**Comment by kovidgoyal** [comment]
This isnt doable in the general case. The terminal protocol has many instances of query/response. And you cannot have input from multiple sources multiplexed into one terminal pipe. Just imagine what would happen if you press the X key and your co-worker presses the backspace key at the same time. You could get either nothing or X replacing the previous character. The best you can do is have one source in control and the rest are readonly (shadow) sessions that can simply see hwta is happening but not make any changes. Most terminal multiplxeres such as screen and tmux have support for that.

**Reply by anicolao** [reply]
Keystrokes interleaving is a daily occurrence, particularly in vim where one person can start typing commands into vim the other person doesn't even see and so many disasterous things are possible (e.g. I type 9j but you type dd almost any time after the 9 :).

But the general case doesn't interest me: the end result is a very practical, very usable pair programming system that works 99% of the time and whose fails are very visible to its users.

One possibility I could add for example would be for there to be a query/response protocol between the terminal program and kitty to verify that the instance of kitty it is talking to is still alive. All kitty instances which aren't the 'owner' of the session see the challenge and henceforth stop responding to terminal queries; then many can be connected but only one is the 'owner' responsible for the terminal chatter. Of course if the kitty clients are configured differently all kinds of terrible things can result. But if they are configured the same, it doesn't matter who answers. 

So in such a scheme, each kitty would have some unique identifier, for example the shasum of $HOSTNAME.$$ for the kitty instance. In my bash prompt, I can send some command like terminal-needs-owner and the response can be $HOSTNAME.$$ (and multiple of these might be received). The terminal picks one at random and henceforth starts emitting heartbeat $HOSTNAME.$$ expecting an ACK back. If the ACKs stop coming, it reverts to terminal-needs-owner to acquire a new owner.

On the kitty side, any process that sees heartbeat <notme> stops responding to terminal queries. Any process that sees heartbeat <me> responds to say it is alive and starts responding to terminal queries.

Probably terrible for everyone else, but might work for me. Better solutions appreciated.

**Reply by page-down** [reply]
> ... make it trivial to share shells with other users and be able to work together collaboratively ...

This sounds like a feature that could be considered for kitty daemon mode?
Multiple clients can connect to the same daemon and only one client can operate at the same time.

**Comment by kovidgoyal** [comment]
On Thu, Feb 03, 2022 at 06:26:07PM -0800, page-down wrote:
> > ... make it trivial to share shells with other users and be able to work together collaboratively ...
> 
> This sounds like a feature that could be considered for kitty daemon mode?
> Multiple clients can connect to the same daemon and only one client can operate at the same time.

Yes, this is a standard feature of most terminal multiplexers. You can
shadow a session but only one owner is allowed per session.

**Reply by anicolao** [reply]
I guess I don't know what kitty daemon mode is, but in my setup and in my thinking the terminal emulator is a client. kitty isn't even installed on the server side container that I use for development, it'd have no purpose there as the system has no console. I am old school enough to think of a terminal as what it is - a system on the far end of a serial port that interprets a stream of bytes to render the display, so it's nonsensical to talk of a terminal 'daemon'. 

On the server side, the set of things you can connect to is a list of ptys that is managed by dtach and you connect your current console of choices to the server side via whatever mechanism you like (mac terminal, kitty, whatever). kitty in this setup is very convenient because it has features I have struggled to find in other terminal emulators, like the ability to display images.

The idea that there should be a query/response protocol over this serial connection seems like overengineering to me. The way I thought it was supposed to work was simply a database describing the protocol and a one-way, server-driven control of the display; the client to server side should be reserved for user keystrokes. I understand that it doesn't work this way, and I need to work around it.

**Comment by anicolao** [comment]
I had a different idea about how to go about it. I could add a filter between `dtach` and the terminal using `socat` so that as far as everyone is concerned the server side applications are talking to a single terminal as usual. But inside the filter, any time there is a request/response escape code protocol going on, it only forwards through the first response, and accumulates subsequent responses and validates that they are the same as the first response.

For extra credit, the filter could write a log of any cases where the responses from different kitty windows differ (for example, this would probably be commonplace and not an error if asking for the pixel dimensions of the window and they differed; but a problem if asking for the cell widthxheight and they differed). The log could be used for debugging or displayed to the end users if it contains problem lines.

While not really "generic" this approach knows when it it is working and when it is failing and is a hopefully very small component.

Opinions?

**Comment by anicolao** [comment, marked as answer]
For now it was easier to not fix this and just send the output of the latest tool I am using through `cat` to suppress its queries of the terminal. This workaround was trivial and continues to dodge the bigger issue for a long time.
